### PR TITLE
Paged Attention + Speculative Decoding support

### PR DESCRIPTION
**Pull request by JRosenkranz**
In this PR, you will find:

- Paged LLaMA model support (implementation of llama with paged attention kernels being used) - supporting tensor-parallel and compile with no graph breaks
- KVCacheManager class which has an implementation using Paged Attention
- CacheData and CacheDataLayer classes which have implementations using Paged Attention
- Speculative generation and paged generation functions to demonstrate how the Paged LLaMA model can be used (These functions include timing statistics)
- Minimal set of Paged Attention kernels required to perform paged attention
- A script which performs both paged generation and speculative paged generation
- automated build of kernels if cuda available

Notable Benchmarks:

Single GPU
LLaMA 7B Compiled

Paged Attention
prompt_len: 49
output_len: 100

```
100 tokens in 100 steps

time per token: 0.011055195331573486s
```

Paged Attention + Cudagraph (single prompt only, batch is work in progress)
prompt_len: 49
output_len: 100

```
100 tokens in 100 steps

time per token: 0.010312161445617675
```

Speculative Decoding + Paged Attention
prompt_len: 49
output_len: 100

```
100 tokens in 33 steps

time per token: 0.004933960437774658s
```

Speculative Decoding + Paged Attention + Cudagraph (single prompt only, batch is work in progress)
prompt_len: 49
output_len: 100

```
100 tokens in 33 steps

time per token: 0.0042519521713256835s
```

### CODE REVIEW

File: fms_extras/utils/cache/__init__.py [hunk 1]
```diff
@@ -0,0 +1,330 @@
...
+        )
+
+    def get_layer(self, layer_index: int) -> OutOfPlaceCacheDataLayer:
+        return OutOfPlaceCacheDataLayer(data_layer=self.data[layer_index])
+
+    def is_filled(self) -> bool:
+        return self.data[0] is not None
+
+
+def flatten_batch(inp):
```
Comment (ani300) [inline fms_extras/utils/cache/__init__.py]: this might need a better name/comment explaining that it's used for speculative decoding, it wasn't immediately obvious

File: fms_extras/utils/cache/__init__.py [hunk 2]
```diff
@@ -0,0 +1,330 @@
...
+            list of sequence ids to free
+        """
+        pass
+
+
+KVCache = Tuple[torch.Tensor, torch.Tensor]  # (key cache, value cache)
+
+
+@dataclasses.dataclass
+class OutOfPlaceCacheDataLayer(CacheDataLayer):
```
Comment (ani300) [inline fms_extras/utils/cache/__init__.py]: some comments on what this type of cache is trying to do would be appreciated

File: fms_extras/utils/cache/__init__.py (hunk 1 above)
Comment (ani300) [reply fms_extras/utils/cache/__init__.py]: plus, what is a redundant token? what is sized as input/sized as output?

File: fms_extras/utils/cache/__init__.py [hunk 3]
```diff
@@ -0,0 +1,330 @@
...
+        )
+        batch_offset += len(lineages)
+    return (
+        torch.cat(out),
+        ind_out,
+        torch.tensor(ind_flat, device=ind_out.device, dtype=torch.int32),
+    )
+
+
+def select_inflate_dim(inp, inds, dim=0):
```
Comment (ani300) [inline fms_extras/utils/cache/__init__.py]: what do the dimension letters mean here?

Comment (ani300) [reply fms_extras/utils/cache/__init__.py]: should this go on a speculative utils file instead?

File: fms_extras/utils/cache/expandable.py [hunk 4]
```diff
@@ -0,0 +1,181 @@
+import dataclasses
+from typing import Dict, List, Optional, Tuple, Union
+
+import torch
+
+from fms_extras.utils.cache import CacheDataLayer, CacheDataWithMetadata, KVCacheManager
+
+
+@dataclasses.dataclass
+class InPlaceCacheDataLayer(CacheDataLayer):
```
Comment (ani300) [inline fms_extras/utils/cache/expandable.py]: same here, what is this cache trying to accomplish?

File: fms_extras/utils/cache/__init__.py (hunk 2 above)
Comment (JRosenkranz) [reply fms_extras/utils/cache/__init__.py]: We actually do not need `OutOfPlaceCache` for this PR as we are only using paged attention, might be better to just remove these classes to reduce the scope

File: fms_extras/utils/cache/expandable.py (hunk 4 above)
Comment (JRosenkranz) [reply fms_extras/utils/cache/expandable.py]: Similar comment to the above, this is not being used. We can probably remove it to reduce the scope

File: fms_extras/utils/generation.py [hunk 5]
```diff
@@ -0,0 +1,371 @@
...
+    """
+    # Construct batch(es) and initial inputs
+    if decode_model is None:
+        decode_model = model
+    bsize = len(input_ids)
+    result = input_ids  # [b] n
+    # Build padded batched input tensor
+    max_len = max([seq.size(0) for seq in input_ids])
+    n_pads_init = [max_len - seq.size(0) for seq in input_ids]
+    n_pads = torch.Tensor(n_pads_init).to(device=input_ids[0].device, dtype=torch.int)
```
Comment (ani300) [inline fms_extras/utils/generation.py]: prefer using torch.tensor() instead of torch.Tensor

File: fms_extras/utils/generation.py [hunk 6]
```diff
@@ -0,0 +1,371 @@
...
+            child_sequence_ids_flattened.extend(child_sequence_ids)
+
+        # add n_adds tokens to each candidate
+        cache_data = kv_cache_manager.allocate_tokens(
+            num_tokens_per_sequence, child_sequence_ids_flattened
+        )
+        position_ids = cache_data.compute_position_ids(num_tokens_per_sequence)
+
+        # Get candidate set of speculations
+        adds = speculator.generate_suffixes(embeds, input_ids, threshes, top_k)  # b k h
```
Comment (ani300) [inline fms_extras/utils/generation.py]: adds as a var name doesn't map to anything that is easy to pick up

File: fms_extras/utils/generation.py [hunk 7]
```diff
@@ -0,0 +1,371 @@
...
+            [input_ids.unsqueeze(1).expand(bsize, top_k, 1), adds], dim=-1
+        ).int()  # b k 1+h
+
+        this_flatting = False
+        if flatting:
+            flat_inputs, unflat_indices, flat_indices = flatten_batch(
+                input_ids
+            )  # b', b k 1+h, b'
+            compression = flat_inputs.numel() / input_ids.numel()
+            if compression < 0.75:
```
Comment (ani300) [inline fms_extras/utils/generation.py:173]: why?

**Review by ani300 — COMMENTED** [review]
in general looks good, but all the code surrounding speculative generation needs better comments/variable names, it's just hard to follow through in general

most of the paged attn code probably has similar issues but I'm a lot more familiar with that so it was a lot easier to follow through.

File: fms_extras/models/paged_llama.py [hunk 8]
```diff
@@ -0,0 +1,811 @@
...
+                .reshape(*temp.size())
+            )
+
+            new_sd[new_name] = temp
+
+    return new_sd
+
+
+serialization.register_adapter(_architecture_name, "meta", _rename_weights_to_fms)
+serialization.register_adapter(_architecture_name, "hf", _hf_sd_to_fms_sd)
```
Comment (nairbv) [inline fms_extras/models/paged_llama.py:844]: if the adapters are the same maybe we can just register_adapter(name, source, get_adapter(original_name, source)) and avoid the extra (copy/paste?) adapter functions

Comment (JRosenkranz) [reply fms_extras/models/paged_llama.py:844]: They are slightly different because there is now a headless model for PagedLLaMA so the mapping required an extra `headless_model` prefix. The logic is basically all the same though.

File: fms_extras/utils/generation.py (hunk 5 above)
Comment (JRosenkranz) [reply fms_extras/utils/generation.py]: fixed in latest commit

File: fms_extras/utils/cache/__init__.py (hunk 1 above)
Comment (daviswer) [reply fms_extras/utils/cache/__init__.py]: This is the function that takes a tree of suffixes:
```
a b c
a b d
a e f
```
and returns `abcdef`. Tokens `a b` in the second line and token `a` in the third line are redundant because we can simply point the attention layer to the corresponding tokens in line 1, which will have the same values. 

I'll make this clearer in the docstring, but this probably does belong as a speculator utility

File: fms_extras/utils/cache/__init__.py (hunk 3 above)
Comment (daviswer) [reply fms_extras/utils/cache/__init__.py]: This is a batched index select that respects the shape of the index tensor. So if index tensor is size (a, ..., z) and dim=0, we'll select the index values from input axis 0, and return the contents in shape (a, ..., z, ...) with trailing dimensions coming from the trailing dimensions of the input tensor. I'll make this clearer (use of x was probably misleading), but like above, it probably does belong as a speculator utility

File: fms_extras/utils/generation.py (hunk 7 above)
Comment (daviswer) [reply fms_extras/utils/generation.py:173]: Flatting the candidate tree introduces some extra overhead in the forward pass, as we have to unflat the batch before each paged attention call. Flatting only becomes worth it at roughly 25% compression - it's just a heuristic we baked in. @JRosenkranz we should probably combine this with the `flatting` flag: rather than set True or False, set to a compression threshold. 1 is always compress, 0 is never. Then default it to .75

File: fms_extras/models/paged_llama.py (hunk 8 above)
Comment (JRosenkranz) [reply fms_extras/models/paged_llama.py:844]: Aside from this, I am wondering if it would be valuable to add a `SimpleMapping` adapter that will take in a list of tuples - replacements. This way, rather than re-implementing an adapter, we just do something like:

```python
meta_replacements = [
        (r"^lm_head.weight", "shared.head.weight"),
        ...
        (r"post_attention_layernorm", "ff_ln"),
    ]

serialization.register_adapter("llama", "meta", SimpleMapping(meta_replacements))
```

It saves a small amount of duplication, and may be simpler for people to do

File: fms_extras/utils/generation.py [hunk 9]
```diff
@@ -125,43 +130,45 @@ def speculative_generate(
...
-        # add n_adds tokens to each candidate
+        # add inp_len tokens to each candidate
         cache_data = kv_cache_manager.allocate_tokens(
             num_tokens_per_sequence, child_sequence_ids_flattened
         )
         position_ids = cache_data.compute_position_ids(num_tokens_per_sequence)
 
         # Get candidate set of speculations
-        adds = speculator.generate_suffixes(embeds, input_ids, threshes, top_k)  # b k h
+        input_ids = speculator.generate_suffixes(embeds, input_ids, threshes, n_candidates)  # b k h
```
Comment (JRosenkranz) [inline fms_extras/utils/generation.py]: I see you are setting the `generate_suffixes` to input_ids then calling cat with the same input_ids as before, overwriting here will not cause issues on the next cat? Does ths assume adds will always be the same as the original input_ids?

**Review by JRosenkranz — COMMENTED** [review]
comment regarding generate_suffixes call

File: fms_extras/utils/generation.py (hunk 9 above)
Comment (daviswer) [reply fms_extras/utils/generation.py]: ~~Previously, we created `adds` and then immediately concatenated to `input_ids` and assigned that to `input_ids`. Then `adds` never got used again. So this is cleaner, and `input_ids` was getting overwritten anyways~~

Comment (daviswer) [reply fms_extras/utils/generation.py]: oh nvm I see what you're saying

File: fms_extras/models/paged_llama.py [hunk 10]
```diff
@@ -0,0 +1,844 @@
...
+# params emb_dim heads layers lr
+#  7B    4096    32    32     3.0E-04
+# 13B    5120    40    40     3.0E-04
+# 33B    6656    52    60     1.5.E-04
+# 65B    8192    64    80     1.5.E-04
+
+
+@dataclass
+class PagedLLaMAConfig(ModelConfig):
+    src_vocab_size: int = 32_000  # can be set by tokenizer
```
Comment (nairbv) [inline fms_extras/models/paged_llama.py:49]: I think the comment isn't clear here, a don't think using a tokenizer will lead to this getting updated automatically right?

File: fms_extras/models/paged_llama.py (hunk 8 above)
Comment (nairbv) [reply fms_extras/models/paged_llama.py:844]: that might be nice if it works, though we do have some transposes and such here too so not sure when it would be useful.

I'm not loving having so much code copy/pasted from the main repo with small tweaks but I guess this is supposed to be temporary until we implement PagedTensor?

Another difference is that we're using fused weights here, which we haven't landed in the main repo yet either.
